### ISSUE TITLE
On some systems, Clinfo reports twice as many platforms as actually exist.

### ISSUE THREAD

**Issue by JStateson** [Oblomov/clinfo#85]
On some systems with mixed AMD co-processors, Clinfo reports twice the number of platforms and twice the number of co-processors.   I suspect this is not a bug in Clinfo but a misconfiguring of the windows AMD drivers that affects only OpenCL.  For example, both the device manager and the AMD Radeon manager correctly list only 5 graphic cards, but Clinfo reports 10 co-processors instead of just 5.  Alternately, there is a deficiency in Clinfo and other OpenCl apps in the method used to obtain the platform info (just guessing).  In any event it adversely affects applications that assign tasks to the coprocessors

![dm-5](https://user-images.githubusercontent.com/37131659/214339876-9cabbfb5-5ba2-498b-8dfc-62ee7ebb14b6.jpg)

![radeon-5](https://user-images.githubusercontent.com/37131659/214340184-f1833fb5-af46-47f8-b250-867793e509cd.jpg)

`C:\Users\josep\Documents>find /i "number of devices" clinfo.txt
---------- CLINFO.TXT
Number of devices:                               5
Number of devices:                               5

NULL platform behavior
  clGetPlatformInfo(NULL, CL_PLATFORM_NAME, ...)  No platform
  clGetDeviceIDs(NULL, CL_DEVICE_TYPE_ALL, ...)   No platform
  clCreateContext(NULL, ...) [default]            No platform
  clCreateContext(NULL, ...) [other]              Success [AMD]
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_DEFAULT)  Success (1)
    Platform Name                                 AMD Accelerated Parallel Processing
    Device Name                                   Hawaii
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_CPU)  No devices found in platform
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_GPU)  Success (5)
    Platform Name                                 AMD Accelerated Parallel Processing
    Device Name                                   Hawaii
    Device Name                                   Hawaii
    Device Name                                   Hawaii
    Device Name                                   Hawaii
    Device Name                                   gfx906
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_ACCELERATOR)  No devices found in platform
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_CUSTOM)  No devices found in platform
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_ALL)  Success (5)
    Platform Name                                 AMD Accelerated Parallel Processing
    Device Name                                   Hawaii
    Device Name                                   Hawaii
    Device Name                                   Hawaii
    Device Name                                   Hawaii
    Device Name                                   gfx906
`

**Comment by JStateson** [comment]
https://stateson.net/images/clinfo.txt

If I can help debug this, let me know.  
Could the problem be multiple libraries due to a driver upgrade?

**Comment by Oblomov** [comment]
OK, it looks like you have multiple platforms defined. You seem to be on Windows, so your best bet would be to check the `HKLM\SOFTWARE\Khronos\OpenCL\Vendors` and/or `HKLM\SOFTWARE\WOW6432Node\Khronos\OpenCL\Vendors` registry keys to see if there's some double definition. It's possible that the driver upgrade caused it, in which case you should probably report this to AMD.

I'm closing this since this is most likely not a clinfo issue, but feel free to reopen if you have more data that would indicate otherwise.

**Comment by Oblomov** [comment]
BTW the clinfo output you provided seems to be from AMD's own clinfo, not this one, although the `clinfo.txt` earlier looks more like the output of this one.

**Comment by JStateson** [comment]
Sorry, my mistake - posted the AMD one.
The correct one is here.
https://stateson.net/images/clinfo_new.txt
I did not find either of those two keys you mentioned.,

This problem has shown up both in Ubuntu as well as Windows and with Nvidia cards, not just AMD cards.  If I can find out how Clinfo somehow gets extra cards, I might be able to spot the problem with the BOINC app that has the same result.  Usually, the problem in BOINC is fixed by using a DDU or uninstaller like Revo but that did not work on my system after several tries.
https://boinc.berkeley.edu/forum_thread.php?id=12807

A solution that some users have found is to use a DDU to remove all display drives, then remove the ethernet cable before rebooting to prevent windows from installing a driver.  The correct driver is then installed and the ethernet re-connected.
This is a workaround.  Best is to find out how the extra Opencl are generated in the first place.

The problem I saw once with Nvidia cards affected only Opencl and not CUDA.

Thank you for responding.

**Comment by Oblomov** [comment]
Rather than extra devices, the issue seems to be that the OpenCL library (the ICD loader) is finding multiple copies of the same platform. Unless you have a custom ICD loader, the list of registered platforms is found in those keys in Windows, and under `/etc/OpenCL/vendors` in Linux and BSD. I have never used BOINC, but it's possible it's somehow interfering with this process, possibly by overriding the ICD loader in some way.

**Comment by JStateson** [comment]
I never found how the ICD loader works in windows.  Those keys you mentioned are not installed by the AMD drivers.  However, I did manage a workaround.  I reverted from AMD 21q2.1 to 21q1.1 and that removed the extra or "ghost" GPUs.  The install of 21q1.1 asked if I wanted to disable windows update during the install in addition to cleaning all the previous AMD drivers.  Possibly that made the difference.  When running 21q2.1  the apps GPUz and MSI afterburner were unable to report many sensor reading.  With 21q1.1 the temps, clock speeds and utilization are all reported correctly and agree with the AMD manager readings.   Your CLinfo now reports the correct number of GPUs.
